spec/unit/ace/client/catalog_spec.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'webmock/rspec'
+require 'ace/client/catalog'
+require 'ace/config'
+
+RSpec.describe ACE::Client::Catalog do
+  describe '#new' do
+    context 'when a valid config is passed' do
+      let(:config) { ACE::Config.new({}) }
+
+      it { expect { described_class.new(config) }.not_to raise_error }
+    end
+
+    context 'when an invalid config is passed' do
+      let(:config) { {} }
+
+      it { expect { described_class.new(config) }.to raise_error ACE::Error, /`config` must be an ACE::Config/ }
+    end
+  end
+
+  describe '#retrieve' do
+    let(:config) { ACE::Config.new(config_data) }
+    let(:instance) { described_class.new(config) }
+    let(:config_data) do
+      {
+        'ssl-cert' => 'spec/fixtures/ssl/cert.pem',
+        'ssl-key' => 'spec/fixtures/ssl/key.pem',
+        'ssl-ca-cert' => 'spec/fixtures/ssl/ca.pem',
+        'puppet-server-uri' => 'https://foo:1234'
+      }
+    end
+    let(:request) do
+      {
+        certname: 'foo.example.com',
+        environment: 'development',
+        facts: {},
+        trusted: {},
+        transaction_uuid: '<transaction_id>',
+        job_id: '<job_id>'
+      }
+    end
+    let(:request_body) do
+      {
+        "certname": 'foo.example.com',
+        "persistence": {
+          "facts": true, "catalog": true
+        },
+        "environment": 'development',
+        "facts": {
+          "values": {}
+        },
+        "trusted_facts": {
+          "values": {}
+        },
+        "transaction_uuid": '<transaction_id>',
+        "job_id": '<job_id>',
+        "options": {
+          "prefer_requested_environment": true,
+          "capture_logs": false
+        }
+      }
+    end
+
+    it 'config should be valid' do
+      expect { config.validate }.not_to raise_error
+    end
+
+    def stub_catalog_request(**options)
```

Comment:
put utility methods at the very top of the context to make them easier to find

Reply:
This is near the top just after the describe